python/cudf/cudf/core/dataframe.py

```diff
@@ -1820,19 +1820,29 @@ def _make_operands_and_index_for_binop(
             NotImplementedType,
         ],
         Optional[BaseIndex],
+        bool,
     ]:
         lhs, rhs = self._data, other
         index = self._index
         fill_requires_key = False
         left_default: Any = False
+        equal_columns = False
+        can_use_self_column_name = True
 
         if _is_scalar_or_zero_d_array(other):
             rhs = {name: other for name in self._data}
+            equal_columns = True
         elif isinstance(other, Series):
-            rhs = dict(zip(other.index.values_host, other.values_host))
+            rhs = dict(zip(other.index.to_pandas(), other.values_host))
             # For keys in right but not left, perform binops between NaN (not
             # NULL!) and the right value (result is NaN).
             left_default = as_column(np.nan, length=len(self))
+            equal_columns = other.index.to_pandas().equals(
+                self._data.to_pandas_index()
+            )
+            can_use_self_column_name = equal_columns or (
+                list(other._index._data.names) == self._data._level_names
+            )
```

Comment:
Can we do the cheap thing first So that we don't convert to pandas and check equality if we don't need to?

Reply:
We could change this to something like this:
```python
if cudf.get_option("mode.pandas_compatible"):
    equal_columns = other.index.to_pandas().equals(self._data.to_pandas_index())
else:
    can_use_self_column_name = (list(other._index._data.names) == self._data._level_names)
    if not can_use_self_column_name:
        can_use_self_column_name = other.index.to_pandas().equals(self._data.to_pandas_index())
```

Reply:
Switched to a cheaper condition first.